cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/io/cdap/cdap/datapipeline/DataPipelineTest.java

```diff
@@ -181,6 +198,102 @@ public void cleanupTest() {
     getMetricsManager().resetAll();
   }
 
+  @Test
+  public void testActionFieldLineage() throws Exception {
+    testActionFieldLineage(Engine.MAPREDUCE);
+    testActionFieldLineage(Engine.SPARK);
+  }
+
+  private void testActionFieldLineage(Engine engine) throws Exception {
+    String readDataset = "ActionReadDataset" + engine;
+    String writeDataset = "ActionWriteDataset" + engine;
+    List<String> srcFields = ImmutableList.of("srcField1", "srcField2", "srcField3");
+    Set<String> destFields = ImmutableSet.of("destField1", "destField2", "destField3");
+
+    List<Operation> operations = new ArrayList<>();
+    /*
+     *          |---------> srcField1 -> destField1----|
+     *          |                                      |
+     * ActionReadDataset -> srcField2 -> destField2 ---|-> ActionWriteDataset
+     *          |                                      |
+     *          |---------> srcField3 -> destField3 ---|
+     */
+    operations.add(new ReadOperation("Read", "1st operation", EndPoint.of("default", readDataset), srcFields));
+    operations.add(new TransformOperation("Transform1", "2nd operation",
+                                          Collections.singletonList(InputField.of("Read", "srcField1")),
+                                          "destField1"));
+    operations.add(new TransformOperation("Transform2", "3rd operation",
+                                          Collections.singletonList(InputField.of("Read", "srcField2")),
+                                          "destField2"));
+    operations.add(new TransformOperation("Transform3", "4th operation",
+                                          Collections.singletonList(InputField.of("Read", "srcField3")),
+                                          "destField3"));
+    operations.add(new WriteOperation("Write", "5th operation", EndPoint.of("default", writeDataset),
+                                      ImmutableList.of(InputField.of("Transform1", "destField1"),
+                                                       InputField.of("Transform2", "destField2"),
+                                                       InputField.of("Transform3", "destField3"))));
+    ETLStage action = new ETLStage("action", FieldLineageAction.getPlugin(readDataset, writeDataset,
+                                                                           new Gson().toJson(operations)));
+
+    ETLBatchConfig etlConfig = ETLBatchConfig.builder().addStage(action).setEngine(engine).build();
+
+    AppRequest<ETLBatchConfig> appRequest = new AppRequest<>(APP_ARTIFACT, etlConfig);
+    ApplicationId appId = NamespaceId.DEFAULT.app("ActionFieldLineage-" + engine);
+    ApplicationManager appManager = deployApplication(appId, appRequest);
+
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
+    workflowManager.waitForRun(ProgramRunStatus.COMPLETED, 5, TimeUnit.MINUTES);
```

Comment:
should use startAndWaitForRun(), which protects against this returning immediately if there is another successful run that wasn't started on the previous line